private/bufpkg/bufimage/bufimageutil/bufimageutil.go

```diff
@@ -60,6 +61,470 @@ func FreeMessageRangeStrings(
 	return s, nil
 }
 
+// ImageFilteredByTypes returns a minimal image containing only the descriptors
+// required to interpret types.
+//
+// Although this returns a new bufimage.Image, it mutates the original image's
+// underlying file's `descriptorpb.FileDescriptorProto` and the old image should
+// not continue to be used.
+func ImageFilteredByTypes(image bufimage.Image, types []string) (bufimage.Image, error) {
+	imageIndex, err := newImageIndexForImage(image)
+	if err != nil {
+		return nil, err
+	}
+	// Check types exist
+	startingDescriptors := make([]protosource.NamedDescriptor, 0, len(types))
+	for _, typeName := range types {
+		descriptor, ok := imageIndex.NameToDescriptor[typeName]
+		if !ok {
+			return nil, fmt.Errorf("not found: %q", typeName)
+		}
+		if image.GetFile(descriptor.File().Path()).IsImport() {
+			return nil, fmt.Errorf("type %q is in an import", typeName)
+		}
```

Comment:
This error makes sense for `buf build`, but it might be confusing for other commands if we ever support the`--type` flag elsewhere.

For example, imagine getting this error after invoking the following:
```sh
$ buf generate --type google.actions.type.DateTimeRange --include-imports
type "google.actions.type.DateTimeRange" is an import
```

With that said, `buf build` includes imports by default, but the default behavior wouldn't make sense either. For example, it'd be weird for this to work:

```sh
$ buf build buf.build/acme/petapis --type google.actions.type.DateTimeRange 
```

Nothing actionable here, but it might be worth flagging this as something we might need to revisit if we ever support `--type` in the commands that act upon imports. It's easier to start with an error anyway.

Reply:
Yes perhaps this should be configurable, especially if we want to extend beyond `buf build`. When building I had the `GetImage` api case in mind where we decided we wouldn't want it. Lets keep it like this and make it configurable when we reason through if it makes sense for other CLI commands.